File: codex-rs/cli/src/bin/codexflow/routing/recovery_report.rs
use anyhow::Context;
use anyhow::Result;
use anyhow::bail;
use serde::Serialize;
use serde_json::Value;
use std::collections::BTreeMap;
use std::path::Path;

const REPORT_SCHEMA: &str = "codexflow.recovery-report.v1";

#[derive(Debug, Serialize, PartialEq, Eq)]
pub(super) struct RecoveryReport {
    schema: &'static str,
    records: usize,
    latest_recorded_at: Option<String>,
    failure_counts: BTreeMap<String, u64>,
    profile_transitions: BTreeMap<String, u64>,
    verification_depth_counts: BTreeMap<String, u64>,
    retry_allowed: u64,
    retry_blocked: u64,
    strategy_changes: u64,
    retrieval_expansions: u64,
    rollback_recommendations: u64,
    human_gates: u64,
    escalations: u64,
}

pub(super) fn build(project_root: &Path, limit: usize) -> Result<RecoveryReport> {
    let events = super::recovery_ledger::history(project_root, limit)?;
    build_from_events(&events)
}

fn build_from_events(events: &[Value]) -> Result<RecoveryReport> {
    let mut failure_counts = BTreeMap::new();
    let mut profile_transitions = BTreeMap::new();
    let mut verification_depth_counts = BTreeMap::new();
    let mut retry_allowed = 0u64;
    let mut retry_blocked = 0u64;
    let mut strategy_changes = 0u64;
    let mut retrieval_expansions = 0u64;
    let mut rollback_recommendations = 0u64;
    let mut human_gates = 0u64;
    let mut escalations = 0u64;

    for (index, event) in events.iter().enumerate() {
        let decision = event
            .get("decision")
            .and_then(Value::as_object)
            .with_context(|| format!("recovery event {} is missing decision object", index + 1))?;
        let failure_class = required_str(decision.get("failure_class"), "failure_class", index)?;
        let current_profile =
            required_profile(decision.get("current_profile"), "current_profile", index)?;
        let next_profile = required_profile(decision.get("next_profile"), "next_profile", index)?;
        let verification_depth = required_str(
            decision.get("verification_depth"),
            "verification_depth",
            index,
        )?;

        increment(&mut failure_counts, failure_class);
        increment(
            &mut profile_transitions,
            &format!("{current_profile}->{next_profile}"),
        );
        increment(&mut verification_depth_counts, verification_depth);

        if required_bool(decision.get("retry_allowed"), "retry_allowed", index)? {
            retry_allowed = retry_allowed.saturating_add(1);
        } else {
            retry_blocked = retry_blocked.saturating_add(1);
        }
        if required_bool(decision.get("strategy_change"), "strategy_change", index)? {
            strategy_changes = strategy_changes.saturating_add(1);
        }
        if required_bool(
            decision.get("additional_retrieval"),
            "additional_retrieval",
            index,
        )? {
            retrieval_expansions = retrieval_expansions.saturating_add(1);
        }
        if required_bool(
            decision.get("rollback_recommended"),
            "rollback_recommended",
            index,
        )? {
            rollback_recommendations = rollback_recommendations.saturating_add(1);
        }
        if required_bool(decision.get("human_approval"), "human_approval", index)? {
            human_gates = human_gates.saturating_add(1);
        }
        if profile_rank(next_profile) > profile_rank(current_profile) {
            escalations = escalations.saturating_add(1);
        }
    }

    let latest_recorded_at = events
        .last()
        .and_then(|event| event.get("recorded_at"))
        .and_then(Value::as_str)
        .map(str::to_string);

    Ok(RecoveryReport {
        schema: REPORT_SCHEMA,
        records: events.len(),
        latest_recorded_at,
        failure_counts,
        profile_transitions,
        verification_depth_counts,
        retry_allowed,
        retry_blocked,
        strategy_changes,
        retrieval_expansions,
        rollback_recommendations,
        human_gates,
        escalations,
    })
}

fn required_str<'a>(value: Option<&'a Value>, field: &str, index: usize) -> Result<&'a str> {
    value
        .and_then(Value::as_str)
        .with_context(|| format!("recovery event {} has invalid {field}", index + 1))
}

fn required_bool(value: Option<&Value>, field: &str, index: usize) -> Result<bool> {
    value
        .and_then(Value::as_bool)
        .with_context(|| format!("recovery event {} has invalid {field}", index + 1))
}

fn required_profile<'a>(value: Option<&'a Value>, field: &str, index: usize) -> Result<&'a str> {
    let profile = required_str(value, field, index)?;
    profile_rank_checked(profile).with_context(|| {
        format!(
            "recovery event {} has invalid {field} {profile:?}",
            index + 1
        )
    })?;
    Ok(profile)
}

fn profile_rank_checked(profile: &str) -> Result<u8> {
    match profile {
        "fast" => Ok(0),
        "balanced" => Ok(1),
        "deep" => Ok(2),
        "critical" => Ok(3),
        _ => bail!("unknown recovery profile {profile:?}"),
    }
}

fn profile_rank(profile: &str) -> u8 {
    profile_rank_checked(profile).expect("profile validated before rank comparison")
}

fn increment(map: &mut BTreeMap<String, u64>, key: &str) {
    let entry = map.entry(key.to_string()).or_default();
    *entry = entry.saturating_add(1);
}

#[cfg(test)]
mod tests {
    use super::*;

    fn event(
        failure: &str,
        current: &str,
        next: &str,
        retry_allowed: bool,
        strategy_change: bool,
        retrieval: bool,
        rollback: bool,
        human: bool,
        verification: &str,
    ) -> Value {
        serde_json::json!({
            "schema": "codexflow.recovery-event.v1",
            "recorded_at": "2026-09-04T00:00:00.000Z",
            "decision": {
                "failure_class": failure,
                "current_profile": current,
                "next_profile": next,
                "retry_allowed": retry_allowed,
                "strategy_change": strategy_change,
                "additional_retrieval": retrieval,
                "rollback_recommended": rollback,
                "human_approval": human,
                "verification_depth": verification,
            }
        })
    }

    #[test]
    fn aggregates_recovery_trajectory_metrics() {
        let events = vec![
            event(
                "test", "fast", "balanced", true, true, false, true, false, "standard",
            ),
            event(
                "permission",
                "balanced",
                "balanced",
                false,
                false,
                false,
                false,
                true,
                "standard",
            ),
        ];
        let report = build_from_events(&events).expect("report");
        assert_eq!(report.records, 2);
        assert_eq!(report.retry_allowed, 1);
        assert_eq!(report.retry_blocked, 1);
        assert_eq!(report.strategy_changes, 1);
        assert_eq!(report.rollback_recommendations, 1);
        assert_eq!(report.human_gates, 1);
        assert_eq!(report.escalations, 1);
        assert_eq!(report.failure_counts.get("test"), Some(&1));
        assert_eq!(report.profile_transitions.get("fast->balanced"), Some(&1));
    }

    #[test]
    fn rejects_invalid_profile_in_persisted_event() {
        let events = vec![event(
            "reasoning",
            "unknown",
            "critical",
            true,
            true,
            false,
            false,
            true,
            "exhaustive",
        )];
        assert!(build_from_events(&events).is_err());
    }
}

File: tools/codexflow/recovery_report.py
#!/usr/bin/env python3
"""Summarize durable CodexFlow recovery trajectories deterministically."""

from __future__ import annotations

import argparse
from collections import Counter
import json
from pathlib import Path
from typing import Any

EVENT_SCHEMA = "codexflow.recovery-event.v1"
REPORT_SCHEMA = "codexflow.recovery-report.v1"
PROFILE_ORDER = {
    "fast": 0,
    "balanced": 1,
    "deep": 2,
    "critical": 3,
}
MAX_LIMIT = 100_000


def load_events(path: Path, limit: int | None = None) -> list[dict[str, Any]]:
    if limit is not None and not 1 <= limit <= MAX_LIMIT:
        raise ValueError(f"limit must be between 1 and {MAX_LIMIT}")
    if not path.exists():
        return []

    events: list[dict[str, Any]] = []
    with path.open("r", encoding="utf-8") as handle:
        for line_number, line in enumerate(handle, 1):
            if not line.strip():
                continue
            try:
                event = json.loads(line)
            except json.JSONDecodeError as exc:
                raise ValueError(f"invalid JSON on line {line_number}: {exc}") from exc
            validate_event(event, line_number)
            events.append(event)

    if limit is not None:
        events = events[-limit:]
    return events


def validate_event(event: Any, line_number: int) -> None:
    if not isinstance(event, dict):
        raise ValueError(f"line {line_number}: recovery event must be an object")
    if event.get("schema") != EVENT_SCHEMA:
        raise ValueError(f"line {line_number}: unsupported recovery event schema")
    if not isinstance(event.get("recorded_at"), str):
        raise ValueError(f"line {line_number}: recorded_at must be a string")
    decision = event.get("decision")
    if not isinstance(decision, dict):
        raise ValueError(f"line {line_number}: decision must be an object")
    required = [
        "failure_class",
        "current_profile",
        "next_profile",
        "retry_allowed",
        "strategy_change",
        "additional_retrieval",
        "rollback_recommended",
        "human_approval",
        "verification_depth",
    ]
    missing = [key for key in required if key not in decision]
    if missing:
        raise ValueError(f"line {line_number}: missing decision fields {', '.join(missing)}")
    for profile_key in ["current_profile", "next_profile"]:
        profile = decision[profile_key]
        if profile not in PROFILE_ORDER:
            raise ValueError(f"line {line_number}: invalid {profile_key} {profile!r}")


def build_report(events: list[dict[str, Any]]) -> dict[str, Any]:
    failure_counts: Counter[str] = Counter()
    transition_counts: Counter[str] = Counter()
    verification_counts: Counter[str] = Counter()
    retry_allowed = 0
    retry_blocked = 0
    strategy_changes = 0
    retrieval_expansions = 0
    rollback_recommendations = 0
    human_gates = 0
    escalations = 0

    for event in events:
        decision = event["decision"]
        failure_counts[str(decision["failure_class"])] += 1
        current = str(decision["current_profile"])
        next_profile = str(decision["next_profile"])
        transition_counts[f"{current}->{next_profile}"] += 1
        verification_counts[str(decision["verification_depth"])] += 1

        if bool(decision["retry_allowed"]):
            retry_allowed += 1
        else:
            retry_blocked += 1
        strategy_changes += int(bool(decision["strategy_change"]))
        retrieval_expansions += int(bool(decision["additional_retrieval"]))
        rollback_recommendations += int(bool(decision["rollback_recommended"]))
        human_gates += int(bool(decision["human_approval"]))
        escalations += int(PROFILE_ORDER[next_profile] > PROFILE_ORDER[current])

    latest_recorded_at = events[-1]["recorded_at"] if events else None
    return {
        "schema": REPORT_SCHEMA,
        "records": len(events),
        "latest_recorded_at": latest_recorded_at,
        "failure_counts": dict(sorted(failure_counts.items())),
        "profile_transitions": dict(sorted(transition_counts.items())),
        "verification_depth_counts": dict(sorted(verification_counts.items())),
        "retry_allowed": retry_allowed,
        "retry_blocked": retry_blocked,
        "strategy_changes": strategy_changes,
        "retrieval_expansions": retrieval_expansions,
        "rollback_recommendations": rollback_recommendations,
        "human_gates": human_gates,
        "escalations": escalations,
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--project-root",
        type=Path,
        default=Path.cwd(),
        help="Project root containing .codexflow/state/recovery-v1.jsonl",
    )
    parser.add_argument("--limit", type=int, default=None)
    args = parser.parse_args()

    path = args.project_root.resolve() / ".codexflow" / "state" / "recovery-v1.jsonl"
    events = load_events(path, args.limit)
    print(json.dumps(build_report(events), indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
